File: Applied Course/4.Problem Solving/5.Problems on Trees/34.Populating Next Right Pointers in Each Node.py
# https://leetcode.com/problems/populating-next-right-pointers-in-each-node
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return root
        queue=[root,None]
        while queue:
            current=queue.pop(0)
            if current==None:
                if queue:
                    queue.append(None)
            else:
                current.next=queue[0]
                if current.left:
                    queue.append(current.left)
                if current.right:
                    queue.append(current.right)
        return root

File: Applied Course/4.Problem Solving/5.Problems on Trees/34.Populating Next Right Pointers in Each Node.cpp
#include<bits/stdc++.h>
using namespace std;
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* left;
    Node* right;
    Node* next;

    Node() : val(0), left(NULL), right(NULL), next(NULL) {}

    Node(int _val) : val(_val), left(NULL), right(NULL), next(NULL) {}

    Node(int _val, Node* _left, Node* _right, Node* _next)
        : val(_val), left(_left), right(_right), next(_next) {}
};
*/

class Solution {
public:
    Node* connect(Node* root) {
        if(!root)
            return root;
        queue<Node*> q;
        q.push(root);
        q.push(NULL);
        while(!q.empty())
        {
            Node *current=q.front();
            q.pop();
            if(current==NULL)
            {
                if(!q.empty())
                    q.push(NULL);
            }
            else
            {
                current->next=q.front();
                if(current->left)
                    q.push(current->left);
                if(current->right)
                    q.push(current->right);
            }
        }
        return root;
    }
};
int main()
{
    // https://leetcode.com/problems/populating-next-right-pointers-in-each-node
    return 0;
}